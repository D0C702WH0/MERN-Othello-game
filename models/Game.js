
const mongoose = require('mongoose')
var Schema = mongoose.Schema

const gameSchema = new Schema({
  blackPassCount: { type: Number },
  whitePassCount: { type: Number },
  game: { type: Object }
})

module.exports = mongoose.model('Game', gameSchema)
