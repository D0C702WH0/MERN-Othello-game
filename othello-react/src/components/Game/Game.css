.aside_right span {
  display: inline-block;
  width: 50px;
  height: 50px;
  border-radius: 50%;
}

button {
  margin: 1rem 3rem;
}

section {
  display: flex;
  justify-content: space-evenly;
  align-items: center;
}

.pass {
  display: flex;
  justify-content: center;
  align-items: center;
}

@media only screen and (max-width: 930px) {
  button {
    margin: 0.5rem;
  }
}

@media only screen and (min-width: 0px) and (max-width: 767px) {
  aside {
    position: absolute;
    top: 30rem;
  }

  .aside_left {
    left: 15vw;
  }

  .aside_right span {
    display: none;
  }

  .aside_right {
    right: 15vw;
  }
}
