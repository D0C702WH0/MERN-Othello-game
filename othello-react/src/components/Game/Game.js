import React, { Component } from 'react'
import { Button } from 'react-bootstrap'
import Board from '../Board/Board'
import merge from 'lodash.merge'
import reversi from 'reversi'
import axios from 'axios'
import openSocket from 'socket.io-client'

const uri = 'http://localhost:8080'

const socket = openSocket(uri)

const Reversi = reversi.Game

const url = 'http://localhost:8080/api/game'

export default class Game extends Component {
  state = {
    game: null,
    nextPlayer: '',
    score: null,
    id: null,
    blackPassCount: 0,
    whitePassCount: 0
  }

  /** MERGE THE RESULT FROM DB INTO A NEW GAME INSTANCE
   * In order to get all Game class methods
   * @param res data from the mongo db
   * @returns game instance of data in state
   */
  turnDataInGameInstance = (res) => {
    const newGame = new Reversi()
    const dbGame = res.data[0].game
    merge(newGame, dbGame)
    const id = res.data[0]._id
    return this.setState({
      id,
      nextPlayer: newGame._nextPieceType,
      game: newGame,
      blackPassCount: res.data[0].blackPassCount,
      whitePassCount: res.data[0].whitePassCount
    })
  }

  /**
   * FETCH GAME DATA FROM DB
   * And turn it into Game instance
   */
  getGameData = () => {
    axios.get(url).then(res => {
      if (res.data.length) {
        this.turnDataInGameInstance(res)
      }
      return this.countPoints()
    })
      .catch(err => err)
  }

  componentDidMount () {
    this.getGameData()
    socket.on('gameUpdated', () => this.getGameData())
  }

  /**
   * Listens to the socket events
   */
  onMessage = () => {
    return this.getGameData()
  }

  /**
 * Allows to get the game score
 * set the score state according to countByPieceType() method results
 */
  countPoints = () => {
    const { game } = this.state
    if (game !== null) {
      const score = game.board.countByPieceType()
      return this.setState({ score: score })
    }
  }

  updateGame = () => {
    const { id, game, blackPassCount, whitePassCount } = this.state

    axios.put(`${url}/${id}`, {
      whitePassCount,
      blackPassCount,
      game
    })
    return this.setState({ nextPlayer: game._nextPieceType, blackPassCount, whitePassCount })
  }

  /**
 * Allows to change the pawn status if game move is legit
 * @param {*} x pawn X coordinate
 * @param {*} y pawn Y coordinate
 */
  handleClick = (x, y) => {
    const { game } = this.state
    /* CHECK IF THE MOVE IS LEGAL */
    const report = game.proceed(x, y)
    /* RETURNS IF ILLEGAL MOVE */
    if (!report.isSuccess) {
      return
    }
    this.updateGame()
    return this.countPoints()
  }

  handlePass = async () => {
    let { game, blackPassCount, whitePassCount, nextPlayer } = this.state

    switch (game._nextPieceType) {
      case 'BLACK':
        game._nextPieceType = 'WHITE'
        blackPassCount++
        nextPlayer = 'WHITE'
        break
      case 'WHITE':
        game._nextPieceType = 'BLACK'
        whitePassCount++
        nextPlayer = 'BLACK'
        break
      default:
        break
    }

    if (whitePassCount > 1) {
      alert('BLACK wins')
      this.setState({ whitePassCount: 0, blackPassCount: 0 })
      return this.handleNewGame()
    }

    if (blackPassCount > 1) {
      alert('WHITE wins')
      this.setState({ whitePassCount: 0, blackPassCount: 0 })
      return this.handleNewGame()
    }

    await this.setState({ nextPlayer, whitePassCount: whitePassCount, blackPassCount: blackPassCount })
    this.updateGame()
  }

  /**
 * Allows to reset the game
 */
  handleNewGame = async () => {
    const { id, blackPassCount, whitePassCount } = this.state
    const newGame = new Reversi()
    /* IF NO GAME CREATE A NEW ONE */
    if (id === null) {
      this.setState({
        game: newGame,
        whitePassCount: 0,
        blackPassCount: 0,
        nextPlayer: newGame._nextPieceType,
        score: newGame.board.countByPieceType()
      })
      const res = await axios.post(url, { newGame })
      this.setState({ id: res.data._id, nextPlayer: res.data._nextPieceType, blackPassCount: 0, whitePassCount: 0 })
    }
    /* ELSE CLEAR THE EXISTING ONE */
    axios.put(`${url}/newGame/${id}`, { newGame, whitePassCount: 0, blackPassCount: 0 })
      .then(() => this.setState({ game: newGame, nextPlayer: newGame._nextPieceType, blackPassCount, whitePassCount, score: null }))
      .catch(err => err)
  }

  render () {
    const { nextPlayer, score, game, id } = this.state
    const { handleNewGame, handleClick, handlePass } = this
    return (
      <div className="App" >
        {game !== null && !game.isEnded &&
          <>
            <h1>{`Player: ${nextPlayer}`}</h1>
            <span>
              {
                `WHITE: ${score === null ? 2 : score.WHITE} points 
            Vs BLACK: ${score === null ? 2 : score.BLACK} points`
              }
            </span>
          </>
        }
        {game !== null && game.isEnded && <h2>{`${game.getHighScorer()} player wins !`}</h2>}
        <section>
          {
            game !== null &&
            <Board click={handleClick} board={game._board} />
          }
          <Button
            tabIndex={0}
            variant={id === null ? 'primary' : 'danger'}
            onClick={handleNewGame}>
            {id === null ? 'New game' : 'Reset'}
          </Button>
          <Button onClick={handlePass}>Pass</Button>
        </section>
      </div>
    )
  }
}
