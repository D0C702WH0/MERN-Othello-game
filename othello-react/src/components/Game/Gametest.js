// import React from 'react'
// import TestRenderer from 'react-test-renderer'
// import { configure, mount } from 'enzyme'
// import Adapter from 'enzyme-adapter-react-16'
// import Game from './Game'
// import { Socket } from 'react-socket-io'

// const uri = 'http://localhost:8080'
// const options = { transports: ['websocket'] }

// configure({ adapter: new Adapter() })

// let renderer
// beforeEach(() => {
//   renderer = mount(
//     <Socket
//       uri={uri}
//       options={options}
//     >
//       <Game />
//     </Socket>
//   )
// })

// describe('Board', () => {
//   it('Basic rendering', () => {
//     const renderer = TestRenderer.create(
//       <Game />
//     )
//     const result = renderer.toJSON()
//     expect(result).toMatchSnapshot()
//   })

//   // it('Doesn\'t change player after an illegal click', () => {
//   //   const instance = renderer.instance()
//   //   instance.handleClick(3, 3)
//   //   expect(renderer.state('isBlack')).toBe(true)
//   // })

//   // it('Changes player after a legal click', () => {
//   //   const instance = renderer.instance()
//   //   instance.handleClick(2, 3)
//   //   expect(renderer.state('isBlack')).toBe(false)
//   // })

//   // it('Changes score after a legal click', () => {
//   //   const instance = renderer.instance()
//   //   instance.handleClick(2, 3)
//   //   expect(renderer.state('score').BLACK).toEqual(4)
//   //   expect(renderer.state('score').WHITE).toEqual(1)
//   // })

//   // it('Changes black pawns number after a legal click', () => {
//   //   const instance = renderer.instance()
//   //   instance.handleClick(2, 3)
//   //   renderer.update()
//   //   expect(renderer.find('.black')).toHaveLength(4)
//   // })
// })
