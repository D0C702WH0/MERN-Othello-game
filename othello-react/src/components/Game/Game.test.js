import React from 'react'
import TestRenderer from 'react-test-renderer'
import Game from './Game'
import reversi from 'reversi'

const GameGameInstance = reversi.Game
const game = new GameGameInstance()

const state = {
  id: 'sdlkÃ©&1212',
  game: game,
  nextPlayer: game._nextPieceType,
  score: null,
  blackPassCount: 0,
  whitePassCount: 0
}

let renderer
let instance
beforeEach(() => {
  renderer = TestRenderer.create(
    <Game/>
  )
  instance = renderer.root.instance
  TestRenderer.act(() => {
    instance.setState(
      {
        id: state.id,
        game: state.game,
        nextPlayer: game._nextPieceType,
        score: game._board.countByPieceType(),
        blackPassCount: state.blackPassCount,
        whitePassCount: state.whitePassCount
      }
    )
  })
})

describe('Board', () => {
  it('Basic rendering', () => {
    const result = renderer.toJSON()
    expect(result).toMatchSnapshot()
  })

  it('Doesn\'t change player after an illegal click', () => {
    instance.handleClick(3, 3)
    expect(instance.state.nextPlayer).toBe('BLACK')
  })

  it('Changes initial score after a legal click', () => {
    TestRenderer.act(() => {
      instance.handleClick(2, 3)
    })

    expect(instance.state.score.BLACK).toBe(4)
  })

  // it('Changes score after a legal click', () => {
  //   const instance = renderer.instance()
  //   instance.handleClick(2, 3)
  //   expect(renderer.state('score').BLACK).toEqual(4)
  //   expect(renderer.state('score').WHITE).toEqual(1)
  // })

  // it('Changes black pawns number after a legal click', () => {
  //   const instance = renderer.instance()
  //   instance.handleClick(2, 3)
  //   renderer.update()
  //   expect(renderer.find('.black')).toHaveLength(4)
  // })
})
