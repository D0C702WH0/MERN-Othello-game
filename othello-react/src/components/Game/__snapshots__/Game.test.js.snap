// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board Basic rendering 1`] = `
<div
  className="App"
>
  <h1>
    Player: BLACK
  </h1>
  <span>
    WHITE: 2 points 
            Vs BLACK: 2 points
  </span>
  <section>
    <div
      className="board_container"
    >
      <div
        className="board"
      >
        <div
          className="row"
        >
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn white"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn black"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn black"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn white"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
          <div
            className="square"
            onClick={[Function]}
          >
            <div
              className="pawn undefined"
            />
          </div>
        </div>
      </div>
    </div>
    <button
      className="btn btn-danger"
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
      type="button"
    >
      Reset
    </button>
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Pass
    </button>
  </section>
</div>
`;
